src/ILCompiler.Compiler/src/Compiler/PreinitializationManager.cs

```diff
@@ -48,5 +80,92 @@ private static bool HasEagerConstructorAttribute(TypeDesc type)
             return mdType != null && 
                 mdType.HasCustomAttribute("System.Runtime.CompilerServices", "EagerStaticClassConstructionAttribute");
         }
+
+        public bool IsPreinitialized(MetadataType type)
+        {
+            // If the cctor interpreter is not enabled, no type is preinitialized.
+            if (!_enableInterpreter)
+                return false;
+
+            if (!type.HasStaticConstructor)
+                return false;
+            
+            // The cctor on the Module type is the module constructor. That's special.
+            if (type.IsModuleType)
+                return false;
+
+            if (type.HasInstantiation)
+            {
+                // Generic definitions cannot be preinitialized
+                if (type.IsGenericDefinition)
+                    return false;
+
+                // If the type has a canonical form the runtime type loader could create
+                // a new type sharing code with this one. They need to agree on how
+                // initialization happens. We can't preinitialize runtime-created
+                // generic types at compile time.
+                if (type.ConvertToCanonForm(CanonicalFormKind.Specific)
+                    .IsCanonicalSubtype(CanonicalFormKind.Any))
+                    return false;
+            }
+
+            return GetPreinitializationInfo(type).IsPreinitialized;
+        }
+
+        public void LogStatistics(Logger logger)
+        {
+            if (!_enableInterpreter)
+                return;
+
+            int totalEligibleTypes = 0;
+            int totalPreinitializedTypes = 0;
+
+            if (logger.IsVerbose)
```

Comment:
This is super verbose compared to current verbosity of the verbose log (in release build). Should it be under a separate switch?

Reply:
It's still nothing compared to what the spew looks like if we pass `--singlethreaded` though.

Eventually I would like the logger to have message levels, facilities, etc. Right now it's just a poor man's logger, like the comment in Logger.cs states.